src/hexen/r_data.c

```diff
@@ -535,21 +547,33 @@ void R_InitTrueColormaps(char *current_colormap)
 	}
 	else
 	{
+		byte *const colormap = W_CacheLumpName(current_colormap, PU_STATIC);
+
 		for (c = 0; c < NUMCOLORMAPS; c++)
 		{
 			for (i = 0; i < 256; i++)
 			{
-				r = gamma2table[crispy->gamma][playpal[3 * colormap[c * 256 + i] + 0]] & ~3;
-				g = gamma2table[crispy->gamma][playpal[3 * colormap[c * 256 + i] + 1]] & ~3;
-				b = gamma2table[crispy->gamma][playpal[3 * colormap[c * 256 + i] + 2]] & ~3;
+				if (broken_wite && i == 255)
+				{
```

Comment:
But in this case we are referring to palette index indeed, [this condition](https://github.com/fabiangreffrath/crispy-doom/pull/1145/files#diff-7248a9344cd0897f38a10b7666afa8757d3f10dee1d304d3733fe82ccc50ee62R548) is not for `if (crispy->truecolor)`. I have verified with "broken white" sprite with `COLORMAP` lump's replaced index 255, and without custom lump, everything is fine.

Or better just wait when you will have some time, probably I'm getting something wrong. 🙂

Reply:
Okay, I think we have a different problem here: The truecolor code falsely assumes that the first colormap contains a 1:1 mapping of the palette colors to themselves (i.e. `colormap[i] == i` for `0 <= i <= 255`). However, this doesn't seem to be true for color index 255 in Hexen, and who knows where else.

When rendering patches, we need a "palette index to RGB" conversion (i.e. `byte` -> `pixel_t`) and we just use the first colormap for this. However, in the original code, this never goes through the colormap, it uses palette indices directly. So, I am afraid to get this right for *all* games, we need a separate "palette index to RGB" lookup table that we use instead of `colormap[]` when we used to write `pixel_t dst = colormap[src]`.

Reply:
> **Edit:** testing wads, with and w/o custom `COLORMAP` and "broken white" Fighter's sprite: [white-checks.zip](https://github.com/fabiangreffrath/crispy-doom/files/14117229/white-checks.zip)

Actually, the non-truecolor mode is right here. Compare with chocolate-hexen.

Reply:
> Actually, the non-truecolor mode is right here. Compare with chocolate-hexen.

Just tried. So it does happens with PSPrites (handled via `colormaps` for light levels) and not happens with sphere on finale screens. I can understand this issue in general, but how to proceed then, it's a bit too complicated... 🤔 Another confusion is: finale screens are in RAW format, and we have verified before that they should be in `byte` format. And, something similar you have suggested is already happening or [RAW drawing](https://github.com/fabiangreffrath/crispy-doom/blob/master/src/v_video.c#L912), or most likely I'm totally confused at this point.

Good news: I have overthinked blending calling optimization way too much. It's not even needed. Simple `printf` showing perfectly when drawing function is actually called, otherwise buffered drawing is happening:

<details>

```diff
@@ -360,6 +360,7 @@
     }
 #ifdef CRISPY_TRUECOLOR
     // [crispy] apply true color blending on top of patch drawing functions
+    printf(".");
     for (int y = 0; y < SCREENWIDTH * SCREENHEIGHT; y++)
     {
         I_VideoBuffer[y] = I_BlendDark(I_VideoBuffer[y], BlendTic);
 #endif
 }
```

</details>

And it is called only while fading is happened! I was worry too much about it because: should widescreen assets (21:9), _*and/or*_ high(-er) resolutions, _*and/or*_ possibly uncapped framerate on menu/non-level states will be implemented, all of this will turn finale screens to CPU benchmark stress-test.

My only personal nitpick is a value `4` ([here](https://github.com/fabiangreffrath/crispy-doom/pull/1145/files#diff-471deeafc857e3d3041b7de95e911213dbaf94f18861e457707eb97e525bdf80R167)), it should be macrocized to `BLENDSTEP` or `BlendStep` (not very correct in terms of styling, but we don't need extra `int` or `short` for it). The value itself is a bit magic number, but represents duration of original fade almost perfectly. Should it be lowered or raised, this will lead to too slow or too fast fading.